Added function biweight_midcovariance to stats module in funcs.py script

nkern opened the pull request:
I added a function in the stats module in the func.py script called biweight_midcovariance(), which is a robust and resistant estimator of the covariance matrix for a multidimensional sampling of points. It is documented according to the other biweight_***() functions and has an example included in the doc string.

pllim commented on astropy/stats/funcs.py:
```diff
@@ -1019,6 +1020,77 @@ def biweight_midvariance(a, c=9.0, M=None, axis=None):
+    numpy covariance and the biweight_covariance
+
+    >>> import numpy as np
+    >>> from scipy import stats
```
doctest here should be optional because scipy is optional dependency

mirca commented on astropy/stats/funcs.py:
```diff
@@ -28,7 +28,8 @@
                                                   'scipy.optimize',
-                                                  'scipy.integrate']}
+                                                  'scipy.integrate',
+                                                  'scipy.stats']}
```
I think this is wrong, right? I think you should add 'scipy.stats' under a new key called 'biweight_midcovariance'.

mirca commented on astropy/stats/funcs.py:
```diff
@@ -1019,6 +1020,77 @@ def biweight_midvariance(a, c=9.0, M=None, axis=None):
+    >>> import numpy as np
+    >>> from scipy import stats
+    >>> from astropy.stats import biweight_midcovariance
+    >>> d = np.array([stats.norm.rvs(0,1,1000),stats.norm.rvs(0,3,1000)])
```
need white spaces after commas

mirca commented on astropy/stats/funcs.py:
```diff
@@ -1019,6 +1020,77 @@ def biweight_midvariance(a, c=9.0, M=None, axis=None):
+    mask = np.abs(u) < 1
+    n = mask.sum(axis=1)
+    usub1 = (1-u**2)
+    usub5 = (1-5*u**2)
```
add white spaces between math operators

mirca commented on astropy/stats/funcs.py:
```diff
@@ -1019,6 +1020,77 @@ def biweight_midvariance(a, c=9.0, M=None, axis=None):
+    # now compute nominator and denominators
+    numerator = np.array(map(lambda x: d*usub1**2*x, d*usub1**2)).sum(axis=2)
+    denominator = (usub1*usub5).sum(axis=1)
+    denominator = np.array(map(lambda x: denominator*x, denominator))
```
white spaces between math operators here too

mirca commented on astropy/stats/funcs.py:
```diff
@@ -1019,6 +1020,77 @@ def biweight_midvariance(a, c=9.0, M=None, axis=None):
+    usub5 = (1-5*u**2)
+    usub1[~mask] = 0.0
+
+    # now compute nominator and denominators
```
``nomitador`` -> ``numerator``?

nkern commented:
Great, the above changes have been made, I added a basic unit test to test_funcs.py, added the method to the log, and optimized it using numpy dot products instead of FOR loops (should have written it that way in the first place).

mirca commented:
@nkern Looks like you have not committed your unit tests.

bsipocz commented:
Nor fixed the scipy doctest requirement causing doctest failures just as in the first commit. You need to add a new dictionary entry for ``biweight_midcovariance``, similar to the other ones.

I'm cancelling travis and appveyor.

nkern commented:
Okay, just committed the unittest and change log.
For the doctest, I changed to using numpy instead of scipy so I don't need to add a key in __doctest_requires__ dictionary.

mirca commented on astropy/stats/tests/test_funcs.py:
```diff
@@ -242,6 +242,11 @@ def test_biweight_midvariance_axis_3d():
+    """test biweight_midcovariance method"""
+    d = np.array([np.random.normal(0,1,500), np.random.normal(0,5,500)])
+    cov = funcs.biweight_midcovariance(d)
+    assert_allclose(cov.shape[0], d.shape[0])
```
You don't want only to make sure that the arrays shape match, but also that your implementation of ``biweight_midcovariance`` gives appropriate results.

bsipocz commented on astropy/stats/funcs.py:
```diff
@@ -1057,7 +1057,6 @@ def biweight_midcovariance(a, c=9.0, M=None):
     >>> np_cov = np.cov(d)
     >>> bw_cov = biweight_midcovariance(d)
     >>> print(np.sqrt(np_cov.diagonal()), np.sqrt(bw_cov.diagonal()))
-    (array([ 1.38868655,  2.97004256]), array([ 1.0232835 ,  3.02582728]))
```
some output is needed here otherwise the doctest fail. Just copy xactly what you get when running the code

nkern replied:
Ah, okay. Got it!

mirca commented on astropy/stats/funcs.py:
```diff
@@ -1019,6 +1019,77 @@ def biweight_midvariance(a, c=9.0, M=None, axis=None):
+    >>> bw_cov = biweight_midcovariance(d)
+    >>> print(np.around((np.sqrt(np_cov.diagonal()), np.sqrt(bw_cov.diagonal())), 1))
+    [[ 2.3  3.1]
+     [ 0.9  3.1]]
```
This example looks good, but it would be nicer if you could explain it a little bit more, e. g., any particular reason to set ``d[0,0] = 30.0``?

nkern replied:
Not really any particular reason, just something that is an obvious outlier of what you would expect with a normal distribution with scale = 1 or 3. I can add a comment that explains that.

mirca commented on astropy/stats/tests/test_funcs.py:
```diff
@@ -242,6 +242,16 @@ def test_biweight_midvariance_axis_3d():
 
+def test_biweight_midcovariance():
+    """test biweight_midcovariance method"""
+    d = np.array([np.random.normal(0,1,500), np.random.normal(0,5,500)])
```
white spaces after commas

nkern replied:
Ok

mirca commented on astropy/stats/tests/test_funcs.py:
```diff
@@ -242,6 +242,16 @@ def test_biweight_midvariance_axis_3d():
+    # Check result is **roughly** correct
+    assert_allclose(std[0],1,0.5)
+    assert_allclose(std[1],5,2.5)
 
```
Do you think it's worth to add a test with ``N_dims > 2``?

nkern replied:
Sure, I'll do that.

bsipocz commented on astropy/stats/funcs.py:
```diff
@@ -1019,6 +1019,77 @@ def biweight_midvariance(a, c=9.0, M=None, axis=None):
+
+    For more details, see:
+    `<http://www.itl.nist.gov/div898/software/
+    dataplot/refman2/auxillar/biwmidc.htm>_.`
```
This link cannot be resolved in this form by the docs builder and thus raises a failure in travis. 

Also, I suggest to move the link to the bottom of the docstring under a new ``References`` section. You can find examples for formatting that in the docsstring of other stats functions (e.g. jackknife or lombscargle)

nkern replied:
Ok will do

crawfordsm commented:
Thanks for the contribution @nkern .  It is definitely very welcome!

For the other versions of biweight, we have added the axis keyword.  Can you add that to your code?  

Two tests to add:
- Can you add a test for a small number of values?  
- Can you add a test for using axis?
- And can you add a test for 3d data?

And is there any chance you could add a comment here with a comparison with its results compared to the code you link to in the comment?  Or any other implementations of the algorithm?

nkern commented:
Hi @crawfordsm,

• I was trying to make this covariance estimator follow the same structure as numpy's covariance estimator so as to be compatible with it, in that the 0th axis represents a variable and the 1st axis represents observations of that variable (numpy.cov has no axis keyword argument). I'm also not sure how easy that would be to allow for arrays greater than 2D; I think one would have to specify both the observation axis and the variable axis, which gets a little complicated. I'll add more clarification to the docstring of the input array that follows the numpy.cov explanation. I did add a "transpose" keyword to allow for taking the covariance along the 0th axis instead of the 1st axis (which happens by default)--which is sort of like allowing axis to be 0 instead of 1 by default. 

• I added an extra test with small N_sample

• The formalism that I reference is written in an older language (dataplot) that I've never used (or heard of), however, I did add a test that cross-references the diagonal of my biweight_midcovariance method with the existing biweight_midvariance method and they give the same results, which is reassuring.

crawfordsm commented:
Ah thanks that makes sense.  In that case,  can I ask you to add a check that if it isn't a 2D array it raises an error and a test for that?    That will be my last request and then I'll merge it.  

Thanks for the contribution.

nkern commented:
Great, its taken care of.

bsipocz commented on astropy/stats/tests/test_funcs.py:
```diff
@@ -242,6 +242,39 @@ def test_biweight_midvariance_axis_3d():
+    test that midcovariance raises error when feed 3D array
+    """
+    d = np.arange(27).reshape(3,3,3)
+    assert_raises(ValueError, funcs.biweight_midcovariance, d)
```
We do not require nose for our tests, so please use ``pytest.raises`` instead. There are plenty of examples in this file.

nkern replied:
Got it, thanks for the help

crawfordsm commented:
@bsipocz are you happy with this?  If so, then this can be merged.

bsipocz commented:
@crawfordsm - Indeed, I haven't noticed the fix. Thanks @nkern for the contribution!
